Clarify undefined behavior and notrap.

sunfishcode opened the pull request:
This PR proposes that load/store/etc. become plain native unchecked memory accesses, without reliable trapping, and with full UB on OOB.

Verifiable safety may be had by using a heap_addr/stack_addr/global_addr immediately before every raw access. Bounds-check elimination optimizations may be performed by using a single heap_addr/stack_addr/global_addr for multiple accesses. Load DCE and DSE can be performed on normal load/store/etc. in the usual ways.

stoklund commented:
We need reliable trapping semantics in order to implement a sandbox. The guard pages after a heap are required to cause a trap when accessed. This is part of the load/store semantics when there is no `notrap` flag.

I don't want to copy the C/LLVM model for plain loads and stores where trapping is covered by the undefined behavior, we should explicitly include trapping as part of the load/store semantics.

The question is how does the `notrap` flag change the semantics of loads and stores? What do we actually need?

You don't define what you meant by "OOB". Do you mean in terms of a WebAssembly heap?

stoklund commented:
To be clear, the language in `README.rst` is quite out of date now, and it makes sense to delete the UB blurb.

My old idea of an easily verifiable subset didn't work out either. Sandboxed code today uses `load` and `store` instructions which are not part of a safe subset. You can still verify safety by looking at where the addresses come from, but it is not a s simple as checking for a safe subset of instructions.

It turns out that safety verification isn't really relevant for Cretonne.

sunfishcode commented:
That makes sense; we don't want trapping to be covered by UB. I've now updated the PR to include trapping as part of the possible defined behavior. I also reintroduced `notrap`, modified to produce UB if the memory is not mapped. This feels like what we actually need, though I should think about it some more. We don't need to rush this.

stoklund commented on docs/langref.rst:
```diff
@@ -415,18 +415,30 @@ Indirect function calls use a signature declared in the preamble.
+:inst:`global_addr` with sufficient size to cover the access given its size
+and offset, and there are no intervening memory reallocations, they are *safe*,
+and either succeed or reliably *trap*, depending on whether the computed address
+is *mapped* or not.
```
Instead of using italics, we should define these terms in the glossary and refer to them with `:term:`

sunfishcode replied:
Ok, I've drafted up possible definitions for "mapped", "trap", and "safe". They're still a little vague in some respects, but I think this takes us in the right direction.

stoklund commented on docs/langref.rst:
```diff
@@ -415,18 +415,30 @@ Indirect function calls use a signature declared in the preamble.
+the computed address is :term:`mapped` or not.
+
+When these conditions are not met, the behavior of these instructions is
+undefined.
```
This sentence makes it impossible to map Rust to Cretonne, doesn't it?

stoklund commented:
I think we should avoid defining "safe" as the opposite of "undefined behavior" or vice versa. I think your glossary definition of "safe" is OK in that respect.

We don't want a Rust program to be undefined behavior just because it doesn't use the heaps and globals.

sunfishcode commented:
Yes; I was thinking too much about the WebAssembly use cases there. I've now introduced concepts of "addressable" and "accessible" memory, which seem like a useful vocabulary to describe memory outside of just heaps, globals, and the stack.

stoklund commented:
I think we can land this in the current state. Please squash the commits into one, though.

sunfishcode commented:
Landed. We can obviously continue to refine these concepts and the documentation as needed.
